url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/layout/hierarchical/model/mxGraphHierarchyModel.java

```diff
@@ -0,0 +1,1001 @@
+			Object[] conns = graph.getConnections(vertices[i], parent);
+			List<Object> outgoingCells = Arrays.asList(graph.getOpposites(
+					conns, vertices[i]));
+			internalVertices[i].connectsAsSource = new LinkedHashSet<mxGraphHierarchyEdge>(
+					outgoingCells.size());
+
+			// Create internal edges, but don't do any rank assignment yet
+			// First use the information from the greedy cycle remover to
+			// invert the leftward edges internally
+			Iterator<Object> iter = outgoingCells.iterator();
+
+			while (iter.hasNext())
+			{
+				// Don't add self-loops
+				Object cell = iter.next();
+
+				if (cell != vertices[i] && graph.getModel().isVertex(cell)
+						&& !layout.isVertexIgnored(cell))
+				{
+					// Allow for parallel edges
+					Object[] edges = graph.getEdgesBetween(vertices[i], cell,
+							true);
+
+					if (edges != null && edges.length > 0)
+					{
+						ArrayList<Object> listEdges = new ArrayList<Object>(
+								edges.length);
+
+						for (int j = 0; j < edges.length; j++)
+						{
+							listEdges.add(edges[j]);
+						}
+
+						mxGraphHierarchyEdge internalEdge = new mxGraphHierarchyEdge(
+								listEdges);
+						Iterator<Object> iter2 = listEdges.iterator();
+
+						while (iter2.hasNext())
+						{
+							Object edge = iter2.next();
+							edgeMapper.put(edge, internalEdge);
+
+							// Resets all point on the edge and disables the edge style
+							// without deleting it from the cell style
+							graph.resetEdge(edge);
+
+							if (layout.isDisableEdgeStyle())
+							{
+								layout.setEdgeStyleEnabled(edge, false);
+								layout.setOrthogonalEdge(edge,true);
+							}
+						}
+
+						internalEdge.source = internalVertices[i];
+						internalVertices[i].connectsAsSource.add(internalEdge);
+					}
+				}
+			}
+
+			// Ensure temp variable is cleared from any previous use
+			internalVertices[i].temp[0] = 0;
+		}
+	}
+
+	/**
+	 * Basic determination of minimum layer ranking by working from from sources
+	 * or sinks and working through each node in the relevant edge direction.
+	 * Starting at the sinks is basically a longest path layering algorithm.
+	 */
+	public void initialRank()
+	{
+		Collection<mxGraphHierarchyNode> internalNodes = vertexMapper.values();
+		LinkedList<mxGraphHierarchyNode> startNodes = new LinkedList<mxGraphHierarchyNode>();
+
+		if (!scanRanksFromSinks && roots != null)
+		{
+			Iterator<Object> iter = roots.iterator();
+
+			while (iter.hasNext())
+			{
+				mxGraphHierarchyNode internalNode = vertexMapper.get(iter
+						.next());
+
+				if (internalNode != null)
+				{
+					startNodes.add(internalNode);
+				}
+			}
+		}
+
+		if (scanRanksFromSinks)
+		{
+			Iterator<mxGraphHierarchyNode> iter = internalNodes.iterator();
+
+			while (iter.hasNext())
+			{
+				mxGraphHierarchyNode internalNode = iter.next();
+
+				if (internalNode.connectsAsSource == null
+						|| internalNode.connectsAsSource.isEmpty())
+				{
+					startNodes.add(internalNode);
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 402:415, 418:431.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.